packages/cli/src/utils/build.js

```diff
@@ -71,171 +70,396 @@ const requiredFiles = async ({ cwd, entryPoints }) => {
-  return new Promise((resolve, reject) => {
-    const output = fs.createWriteStream(zipPath);
-    const zip = archiver('zip', {
-      zlib: { level: Z_BEST_COMPRESSION },
-    });
+const looksLikeWorkspaceRoot = async (dir) => {
+  if (fs.existsSync(path.join(dir, 'pnpm-workspace.yaml'))) {
+    return true;
+  }
 
-    // listen for all archive data to be written
-    output.on('close', function () {
-      resolve();
-    });
+  const packageJsonPath = path.join(dir, 'package.json');
+  if (!fs.existsSync(packageJsonPath)) {
+    return false;
+  }
 
-    zip.on('error', function (err) {
-      reject(err);
-    });
+  let packageJson;
+  try {
+    packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
+  } catch (err) {
+    return false;
+  }
 
-    // pipe archive data to the file
-    zip.pipe(output);
+  return packageJson?.workspaces != null;
+};
 
-    paths.forEach(function (filePath) {
-      let basePath = path.dirname(filePath);
-      if (basePath === '.') {
-        basePath = undefined;
-      }
-      const name = path.join(dir, filePath);
-      zip.file(name, { name: filePath, mode: 0o755 });
-    });
+// Traverses up the directory tree to find the workspace root. The workspace
+// root directory either contains pnpm-workspace.yaml or a package.json file
+// with a "workspaces" field. Returns the absolute path to the workspace root
+// directory, or null if not found.
+const findWorkspaceRoot = async (workingDir) => {
+  let dir = workingDir;
+  for (let i = 0; i < 500; i++) {
+    if (await looksLikeWorkspaceRoot(dir)) {
+      return dir;
+    }
+    if (dir === '/' || dir.match(/^[a-z]:\\$/i)) {
+      break;
+    }
+    dir = path.dirname(dir);
+  }
+  return null;
+};
 
-    zip.finalize();
-  });
+const getNearestNodeModulesDir = (workingDir, relPath) => {
+  if (path.basename(relPath) === 'package.json') {
+    const nmDir = path.resolve(
+      workingDir,
+      path.dirname(relPath),
+      'node_modules',
+    );
+    return fs.existsSync(nmDir) ? path.relative(workingDir, nmDir) : null;
+  } else {
+    let dir = path.dirname(relPath);
+    for (let i = 0; i < 100; i++) {
+      if (dir.endsWith(`${path.sep}node_modules`)) {
+        return dir;
+      }
+      const nextDir = path.dirname(dir);
+      if (nextDir === dir) {
+        break;
+      }
+      dir = nextDir;
+    }
+    return null;
+  }
 };
 
-const makeZip = async (dir, zipPath, disableDependencyDetection) => {
-  const entryPoints = [path.resolve(dir, 'zapierwrapper.js')];
+const writeBuildZipDumbly = async (workingDir, zip) => {
+  for (const entry of walkDirWithPresetBlocklist(workingDir)) {
+    const absPath = path.resolve(entry.parentPath, entry.name);
+    const relPath = path.relative(workingDir, absPath);
+    if (entry.isFile()) {
+      zip.file(absPath, { name: relPath });
+    } else if (entry.isSymbolicLink()) {
+      const target = path.relative(entry.parentPath, fs.realpathSync(absPath));
```

Comment:
```suggestion
      const target = path.relative(entry.parentPath, await fs.realpath(absPath));
```
For non-blocking IO performance. It would be good to wrap it in try...catch to handle errors.

Reply:
@kola-er do you mean wrap the `realpath` call or the entire function body in `try ... catch`? And what error(s) do you expect us to handle?